skimage/filters/rank/core_cy.pyx

```diff
@@ -271,3 +286,293 @@ cdef void _core(void kernel(dtype_t_out*, Py_ssize_t, Py_ssize_t*, double,
+    cdef Py_ssize_t odepth = out.shape[3]
+
+    cdef Py_ssize_t centre_p = <Py_ssize_t>(selem.shape[0] / 2) + shift_x
+    cdef Py_ssize_t centre_r = <Py_ssize_t>(selem.shape[1] / 2) + shift_y
+    cdef Py_ssize_t centre_c = <Py_ssize_t>(selem.shape[2] / 2) + shift_z
+
+    # check that structuring element center is inside the element bounding box
+    assert centre_r >= 0
+    assert centre_c >= 0
+    assert centre_p >= 0
+    assert centre_r < srows
+    assert centre_c < scols
+    assert centre_p < splanes
+
+    cdef Py_ssize_t mid_bin = n_bins / 2
+
+    # define pointers to the data
+    cdef char* mask_data = &mask[0, 0, 0]
+
+    # define local variable types
+    cdef Py_ssize_t r, c, p, j, rr, cc, pp, value, local_max, i, even_row
+
+    # number of pixels actually inside the neighborhood (double)
+    cdef double pop = 0
+
+    # build attack and release borders by using difference along axis
+    t = np.dstack((selem, np.zeros((selem.shape[0], selem.shape[1], 1))))
+    cdef unsigned char[:, :, :] t_e = (np.diff(t, axis=2) < 0).view(np.uint8)
+
+    t = np.dstack((np.zeros((selem.shape[0], selem.shape[1], 1)), selem))
+    cdef unsigned char[:, :, :] t_w = (np.diff(t, axis=2) > 0).view(np.uint8)
+
+    t = np.hstack((selem, np.zeros((selem.shape[0], 1, selem.shape[2]))))
+    cdef unsigned char[:, :, :] t_s = (np.diff(t, axis=1) < 0).view(np.uint8)
+
+    t = np.hstack((np.zeros((selem.shape[0], 1, selem.shape[2])), selem))
+    cdef unsigned char[:, :, :] t_n = (np.diff(t, axis=1) > 0).view(np.uint8)
+
+    # the current local histogram distribution
+    cdef Py_ssize_t* histo
+
+    # these lists contain the relative pixel row and column for each of the 4
+    # attack borders east, west, north and south e.g. se_e_r lists the rows of
+    # the east structuring element border
+    cdef Py_ssize_t se_size = srows * scols * splanes * sizeof(Py_ssize_t)
+    cdef Py_ssize_t* se_e_r
+    cdef Py_ssize_t* se_e_c
+    cdef Py_ssize_t* se_e_p
+
+    cdef Py_ssize_t* se_w_r
+    cdef Py_ssize_t* se_w_c
+    cdef Py_ssize_t* se_w_p
+
+    cdef Py_ssize_t* se_n_r
+    cdef Py_ssize_t* se_n_c
+    cdef Py_ssize_t* se_n_p
+
+    cdef Py_ssize_t* se_s_r
+    cdef Py_ssize_t* se_s_c
+    cdef Py_ssize_t* se_s_p
+
+    # number of element in each attack border
+    cdef Py_ssize_t num_se_n, num_se_s, num_se_e, num_se_w, num_se_u, num_se_d
+
+    with nogil:
+
+        se_e_r = <Py_ssize_t*>malloc(se_size)
+        se_e_c = <Py_ssize_t*>malloc(se_size)
+        se_e_p = <Py_ssize_t*>malloc(se_size)
+
+        se_w_r = <Py_ssize_t*>malloc(se_size)
+        se_w_c = <Py_ssize_t*>malloc(se_size)
+        se_w_p = <Py_ssize_t*>malloc(se_size)
+
+        se_n_r = <Py_ssize_t*>malloc(se_size)
+        se_n_c = <Py_ssize_t*>malloc(se_size)
+        se_n_p = <Py_ssize_t*>malloc(se_size)
+
+        se_s_r = <Py_ssize_t*>malloc(se_size)
+        se_s_c = <Py_ssize_t*>malloc(se_size)
+        se_s_p = <Py_ssize_t*>malloc(se_size)
```

Comment:
> It's also unclear to me why these need to be malloc'ed within a gil-free block, rather than stored in a numpy array created outside of the nogil block? I also wonder if it would not be easier to store these elements as an array of shape `(4, 3, se_size)` rather than as separate variables. What do you think about this idea?

I believe we copied this code template from the 2D version above. I can look into this, and should we change both version?

Reply:
As with the other comments, let us focus on improving the 3D version, so that new code added to the library is always better than what came before. The 2D version can be cleaned up in a future pull request if we think it important. Thank you!